src/server/tiering/small_bins.cc

```diff
@@ -0,0 +1,120 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/tiering/small_bins.h"
+
+#include <algorithm>
+#include <optional>
+#include <utility>
+
+#include "absl/base/internal/endian.h"
+#include "base/logging.h"
+#include "core/compact_object.h"
+#include "server/tiering/disk_storage.h"
+
+namespace dfly::tiering {
+
+std::optional<SmallBins::FilledBin> SmallBins::Stash(std::string_view key, std::string_view value) {
+  DCHECK_LT(value.size(), 2_KB);
+
+  // See FlushBin() for format details
+  size_t value_bytes = 8 /* hash */ + 2 /* length */ + value.size();
+
+  std::optional<FilledBin> filled_bin;
+  if (current_bin_bytes_ + value_bytes >= 4_KB) {
+    filled_bin = FlushBin();
+  }
+
+  current_bin_bytes_ += value_bytes;
+  current_bin_[key] = value;
+  return filled_bin;
+}
+
+SmallBins::FilledBin SmallBins::FlushBin() {
+  std::string out;
+  out.resize(current_bin_bytes_ + 1);
+
+  unsigned id = ++last_bin_id_;
+  auto& pending_list = pending_bins_[id];
+
+  char* data = out.data();
+  while (!current_bin_.empty()) {
+    auto node = current_bin_.extract(current_bin_.begin());
+
+    // Each key/value pair is serialized as:
+    auto& key = node.key();
+    auto& value = node.mapped();
+
+    // 1. 8 byte key hash
```

Comment:
you interleave hashes and key/values. 
Is it how it was in tiered_storage? I think we keep all the hashes together. The motivation  behind the separation: to read easily all the hashes from the page when performing defragmentation.

Reply:
the format was different. The hashes were written first, then all the values with their lengths rounded up because we had separate bins. Will change to all hashes being stored first, but keep in mind this is not a final design